### DISCUSSION TITLE
Best strategy to avoid holes in the mesh?

### DISCUSSION THREAD

**Discussion by ramBrain** [simnibs/simnibs#147] [7.Creating head models] [answered]
Hi all,

With a certain MRI recording, `charm` keeps generating a mesh with some small holes, and I presume that this makes DUNEuro fail when trying to generate a leadfield matrix.
![image](https://github.com/simnibs/simnibs/assets/64736900/a9ad77fa-aafd-4c9a-a53a-ad23823f6512)

Any suggestion on how to properly set setting.ini to yield a better result?

Thank you in advance,
Ramtin

**Comment by jduemose** [comment]
Hi, does duneuro give any kind of message when it fails? Also, in which tissue(s) do these holes appear? I suppose we are looking at a surface here? How do the corresponding volume mesh look?

Best,
Jesper

**Reply by ramBrain** [reply]
> does duneuro give any kind of message when it fails? 

Yes, it prompts:

```
Dune reported error: Dune::Exception [findEntityImpl:/home/juan/bst-duneuro/src/duneuro/duneuro/common/edgehopping.hh:84]: coordinate is outside of the grid, or grid is not convex 

status =

    -1


***************************************************************************
** Error: Error during the DUNEuro computation, see logs in the command window.
***************************************************************************
```

>  in which tissue(s) do these holes appear?

For what I can see, CSF and scalp

> I suppose we are looking at a surface here? How do the corresponding volume mesh look?

Indeed, this is surfaces. I am working on brainstorm, and it is not easy to visualize the volume, as it plots over the MRI along orthogonal axes...

**Reply by jduemose** [reply]
Hi Ramtin

I don't know this particular function or what exactly it checks but perhaps I can have a look when I have time. The FEM system uses the volume elements and not the surfaces and actually in charm the surfaces are mostly reconstructed for visualization afaik so they will not necessarily be "nice" (e.g., closed). That being said, if duneuro works for some subjects but not others, then I guess there is something wrong with those meshes. Is it possible for you to share one of the problematic meshes? Also, you could try to run a simulation (e.g., following one of our tutorials) with a problematic mesh to see if simnibs fails as well.

If you have simnibs installed you can use `gmsh` to view the .msh file it creates. There you can toggle volume elements in case you want to give it a try.

**Reply by ramBrain** [reply]
Hi Jesper,

The failure of DUNEuro might not be related to holes in the mesh, since I tried `headreco`, which paradoxically generated a more precise model than `charm` (without holes), but still could not make DUNEuro succeed.

I cannot share complete meshes for privacy reasons, and I am afraid that a defaced mesh might generate further issues in the model computation... If you think this is not the case, I will be happy to send it through!

Many thanks,
Ramtin

**Reply by jduemose** [reply]
If it is a possibility to deface the MRI and then run charm, you could give that a shot, although the resulting mesh might not produce the same error in duneuro. An idea could also be to ask the brainstorm/duneuro team what this error reflects (i.e., what is checked by this function). That might help us track down potential defects in the mesh.

Best,
Jesper

**Reply by ramBrain** [reply, marked as answer]
With `headreco` it eventually worked after removing holes from the scalp with a function in Brainstorm.
